api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java

```diff
@@ -387,18 +406,181 @@ public boolean supportsPropertyName(String gpName) {
+		PatientIdentifierType idType = patientService.getPatientIdentifierTypeByUuid(patientIdentifierTypeUuid);
+		if (idType == null) {
+			throw new APIException(String.format("PatientIdentifierType is missing: UUID %s",
+					patientIdentifierTypeUuid));
+		} else {
+			//Currently method getPatients() does not take into consideration the identifierTypes,
+			//so it needs to be filtered anyway
+			List<Patient> patients = new ArrayList<Patient>();
+			for (Patient p : patientService.getPatients(null, patientIdentifier,
+					Collections.singletonList(idType), true)) {
+				for (PatientIdentifier pi : p.getIdentifiers()) {
+					if (pi.getIdentifierType().equals(idType)) {
+						patients.add(p);
+						break;
+					}
+				}
+			}
+
+			if (CollectionUtils.isEmpty(patients)) {
+				return null;
+			} else {
+				Patient patient = patients.get(0);
+				dao.getSessionFactory().getCurrentSession().refresh(patient);
+				return patient;
+			}
+			//TODO Add error if more than one patient is has that identifier!
 		}
 	}
 
+	private Patient persistImportedMpiPatient(MpiPatient mpiPatient) {
+		// TODO: Fix bug MPI-11 Updating of Locally generated ID in MPI after MPI import not occurring
+		String openMrsIdTypeUuid = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_UUID);
+
+		PatientIdentifierType openMrsIdType = patientService.getPatientIdentifierTypeByUuid(openMrsIdTypeUuid);
+		Patient patient;
+		if (mpiPatient.getPatientIdentifier(openMrsIdType) == null) {
+			PatientIdentifier localId = validateOrGenerateIdentifier(null, null);
+			mpiPatient.addIdentifier(localId);
+		}
+		patient = patientService.savePatient(mpiPatient);
+		return patient;
+	}
+
+	private IdentifierSourceService getIssAndUpdateIdSource() {
+		IdentifierSourceService iss = Context.getService(IdentifierSourceService.class);
+		if (idSource == null) {
+			String idSourceId = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID);
+			if (StringUtils.isBlank(idSourceId)) {
+				throw new APIException("Please set the id of the identifier source to use to generate patient identifiers");
+			}
+			try {
+				idSource = iss.getIdentifierSource(Integer.valueOf(idSourceId));
+				if (idSource == null) {
+					throw new APIException("cannot find identifier source with id:" + idSourceId);
+				}
+			}
+			catch (NumberFormatException e) {
+				throw new APIException("Identifier source id should be a number");
+			}
+		}
+
+		return iss;
+	}
+
+	private Location getIdentifierLocation(Location identifierLocation) {
+		if (identifierLocation == null) {
+			identifierLocation = locationService.getDefaultLocation();
+			if (identifierLocation == null) {
+				throw new APIException("Failed to resolve location to associate to patient identifiers");
+			}
+		}
+
+		// see if there is a primary identifier location further up the chain, use that instead if there is
```

Comment:
What's the point of using a primary location up the chain?

Reply:
I am not sure... it is in the [current master](https://github.com/openmrs/openmrs-module-registrationcore/blob/e7d2d6b773adc462899b7c1bccac87f0e69aa321/api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java#L190) as well though. Added by @approce when he merged his GSOC MPI project... Do you see any point? if not we can delete it?

@mseaton @mogoodrich @djazayeri